packages/compile-solidity/src/compilerSupplier/loadingStrategies/VersionRange.ts

```diff
@@ -72,7 +81,7 @@ class VersionRange {
       .map(solcVersion => {
         if (semver.satisfies(solcVersion, version)) return solcVersion;
       })
-      .filter(solcVersion => solcVersion);
+      .filter((solcVersion): solcVersion is string => !!solcVersion);
```

Comment:
What does `solcVersion is string` do here? How is it different from `.filter((solcVersion: string) => !!solcVersion)`

Reply:
It's a [User-defined type guard](https://www.typescriptlang.org/docs/handbook/advanced-types.html#user-defined-type-guards); TS knows that the `boolean` returned by that function informs the type of `solcVersion`

Reply:
(Since the `.map()` operation will possibly return `undefined`, returning a type predicate like that means that TS won't think the type of `satisfyingVersions` is `string | undefined`)

Reply:
TIL. Interesting syntax. Thanks!